doc/fluid/design/concepts/python_data_feeding.md

```diff
@@ -0,0 +1,110 @@
+# Python Data Feeding
+
+In the former implementation of Paddle Fluid, there are two ways to feed data:
+
+- Use `reader_op` in backend C++ side. This method only supports data feeding from recordio files and random data generators, but supports many kinds of `decorated_readers`. For examples, `double_buffer_reader` uses two threads to achieve better performance: one for time-consuming I/O operations, and the other for `Executor.Run()`. See [C++ Data Feeding](https://github.com/PaddlePaddle/Paddle/blob/develop/doc/fluid/design/concepts/cpp_data_feeding.md) for details.
+
+- Feed data directly using `DataFeeder.feed()` in Python codes. It is more flexible than the first way. Many kinds of preprocessing steps can be performed before feeding using Python or any other languages, instead of adding many uncommon `operators` in C++ side. But this method is less efficient: the program cannot read the next mini-batch data before `Executor.Run()` ends. Moreover, `decorated_readers` such as `double_buffer_reader` cannot be used for better performance.
+
+In this document, we design a Python Data Feeding process combining the efficiency of the first way and the flexibility of the second way. A data queue `PyArrayFeedQueue` is designed to be shared by the Python and C++ side, while Python array is pushed into the queue and `reader_op` in C++ side reads out the data from the queue.
+
+## Design of PyArrayFeedQueue
+`PyArrayFeedQueue` is a blocking queue with a fixed `capacity` and accepts Python array with shapes indicated by `dims`.
+```C++
+class PyArrayFeedQueueHolder;
+
+class PyArrayFeedQueue {
+  friend class PyArrayFeedQueueHolder;
+ private:
+  PyArrayFeedQueue(size_t capacity, const std::vector<framework::DDim>& dims, const Place& place);
+ public:
+  size_t size() const; // Get the current size of the queue
+  size_t capacity() const; // Get the capacity of the queue
+  bool is_full() const;
+  bool is_empty() const;
+  
+  // Convert Python array tuple to std::vector<framework::LoDTensor> and store it.
+  // Block if is_full() == true
+  // Use pybind11::gil_scoped_release to release GIL of Python
+  void push(const pybind11::tuple& array_tuple);
+  
+  // Block if is_empty() == true
+  // Use pybind11::gil_scoped_release to release GIL of Python
+  std::vector<framework::LoDTensor> pop();
+ private:
+  BlockingQueue<std::vector<framework::LoDTensor>> queue_;
```

Comment:
We have more than one `BlockingQueue` in Fluid, so please use a namespace to specify which one you are using.